parley/src/layout/cursor.rs

```diff
@@ -785,48 +711,62 @@ impl Selection {
                 continue;
             };
             let metrics = line.metrics();
             let line_min = metrics.min_coord as f64;
             let line_max = metrics.max_coord as f64;
-            if line_ix == line_start_ix || line_ix == line_end_ix {
-                // We only need to run the expensive logic on the first and
-                // last lines
-                let mut start_x = metrics.offset as f64;
-                let mut cur_x = start_x;
-                for run in line.runs() {
-                    for cluster in run.visual_clusters() {
-                        let advance = cluster.advance() as f64;
-                        if text_range.contains(&(cluster.text_range().start as u32)) {
-                            cur_x += advance;
-                        } else {
-                            if cur_x != start_x {
-                                let width = (cur_x - start_x).max(MIN_RECT_WIDTH);
-                                f(Rect::new(start_x as _, line_min, start_x + width, line_max));
-                            }
-                            cur_x += advance;
-                            start_x = cur_x;
+            //if line_ix == line_start_ix || line_ix == line_end_ix {
+            // We only need to run the expensive logic on the first and
+            // last lines
+            let mut start_x = metrics.offset as f64;
+            let mut cur_x = start_x;
+            for run in line.runs() {
+                for cluster in run.visual_clusters() {
+                    let advance = cluster.advance() as f64;
+                    if text_range.contains(&cluster.text_range().start) {
+                        cur_x += advance;
+                    } else {
+                        if cur_x != start_x {
+                            let width = (cur_x - start_x).max(MIN_RECT_WIDTH);
+                            f(Rect::new(start_x as _, line_min, start_x + width, line_max));
                         }
+                        cur_x += advance;
+                        start_x = cur_x;
                     }
                 }
-                if cur_x != start_x {
-                    let width = (cur_x - start_x).max(MIN_RECT_WIDTH);
-                    f(Rect::new(start_x, line_min, start_x + width, line_max));
-                }
-            } else {
-                let x = metrics.offset as f64;
-                let width = (metrics.advance as f64).max(MIN_RECT_WIDTH);
-                f(Rect::new(x, line_min, x + width, line_max));
             }
+            if cur_x != start_x {
+                let width = (cur_x - start_x).max(MIN_RECT_WIDTH);
+                f(Rect::new(start_x, line_min, start_x + width, line_max));
+            }
+            // } else {
+            //     let x = metrics.offset as f64;
+            //     let width = (metrics.advance as f64).max(MIN_RECT_WIDTH);
+            //     f(Rect::new(x, line_min, x + width, line_max));
+            // }
```

Comment:
Does this belong here anymore?

Reply:
It does, when the loop actually constrains the expensive code to the first and last lines. Restored all of this code.